Adding new Windows module win_disk_management

marqelme opened the pull request:
##### SUMMARY
This change adds new Windows Ansible module win_disk_management.

With the module you can select a disk on the server and manage it (e.g. initializing, partitioning, formatting). To select the disk and to manage it you have several options which are all described in the documentation.

Supported is:

- check mode
- diff mode
- integration tests

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
lib/ansible/modules/windows/win_disk_management.ps1

##### ANSIBLE VERSION

```
ansible 2.4.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/root/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible
  python version = 2.7.6 (default, Oct 26 2016, 20:30:19) [GCC 4.8.4]
```


##### ADDITIONAL INFORMATION
Within the module you have several options to select the disk in Windows and to manage the disk, partition and volume.
```
- name: Select a defined disk and manage it as specified
  win_disk_management:
    # select_disk_by
    partition_style_select: raw
    operational_status: offline
    read_only: true
    number: 1
    # set_disk_by
    partition_style_set: mbr
    file_system: ntfs
    label: database_disk
    allocation_unit_size: 4
    large_frs: true
    short_names: true

- name: Select a defined disk and manage it as specified
  win_disk_management:
    # select_disk_by
    size: 50
    partition_style_select: mbr
    operational_status: online
    read_only: false
    # set_disk_by
    partition_style_set: gpt
    drive_letter: f
    access_path: C:\Test\
    file_system: refs
    label: application_disk
    allocation_unit_size: 64
    integrity_streams: true
```

```
No changes because the module is a new module.
```

Commit:
Added first windows module win_disk_setup

Commit:
Added first windows module win_disk_setup

Commit:
Fixed some errors and information in win_disk_setup.py

Commit:
Added RETURN in documentation

Commit:
Adjusted failures in documentation

Commit:
Adjusted failures in documentation

Commit:
Adjusted failures in documentation

Commit:
Adjusted failures in documentation

gundalow commented on lib/ansible/modules/windows/win_disk_setup.py:
```diff
@@ -0,0 +1,324 @@
+        - Size of the disk to find
+      required: true
+      default: 5
+      aliases: []
```
Empty aliases lines can be removed

marqelme replied:
Have removed the empty alias

gundalow commented on lib/ansible/modules/windows/win_disk_setup.py:
```diff
@@ -0,0 +1,324 @@
+        - Switch to set Integrity Streams option for file system on found disk, only for ReFs file system
+      required: false
+      default: false
+      choices:
```
Please replace with `type: bool`

Same comment applies to a few other options.

dagwieers replied:
So `required: false` should be removed and this replaces `default` and `choices`:
```yaml
type: bool
default: 'no'
```

marqelme replied:
Thanks for the hints, have changed it to type bool and adjusted ps1 file accordingly

Commit:
Adjusted boolean parameters to type bool, accordingly changed PS file and deleted empty aliases

jhawkesworth commented:
I like the idea of this of this module a lot - it would be a great addition to Ansible to have it included, but I have a few suggestions/ thoughts.

I suggest naming it 'win_disk_management' as its similar to what you can do in the Disk Management snapin.  I notice in the examples its called 'win_disk' - not sure about win_disk_setup as 'setup' at the moment is associated with the idea of gathering facts in ansible, so I'd prefer to avoid that name.

I notice in the documentation that the module parameters are all in CamelCase but ansible module parameters are usually in snake_case (lower case with words separated by _) - I think it would be best to change this for consistency (I appreciate this is not the case in powershell but within playbooks I think it would be best to be consistent).

Regarding the parameter names, I am not sure about the Find... and Set... prefixes - it makes it sound like the module isn't idempotent, and it is repetitive.

It would be a bit more work to split the parameters up so that they are in two different groups, but I think it would make it clearer.  Maybe the names below could be better but something like the following perhaps?

I also like units in the module parameter names - it makes them 'self documenting'.  So you could invoke it something like this:

```yaml
win_disk_management: 
  select_disk_by:
    size_gb: 40
    partition_style: raw
    operational_status: Online
  drive_state: 
    drive_letter: F
    allocation_unit_size_kb: 64
    file_system: NTFS
    label: ApplicationDisk
    short_names: no
    integrity_streams: no
```

dagwieers commented:
@jhawkesworth The "unit in the parameter name" is not the standard way of doing it in Ansible though. I know @nitzmahone prefers this as well, but maybe this should be brought up in a Core meeting before making this a thing ;-)

marqelme commented:
Thanks, the idea to separate the select and set parameters in groups sounds good.
@dagwieers  Should I implement this?
Should I change the name to win_disk_management'?

Commit:
Added unit GB in description for FindSize, corrected module name in examples

dagwieers commented:
I am not convinced that the selection should be part of this module. What we currently lack is proper storage device enumeration (as Windows facts). If we would have that, together with a unique reference (UUID?) to those devices, it would be easy to select a storage device and provide this unique id as input to a module that understands the desired state.

So my advice here is to focus on storage device enumeration (the Windows way) and add it to the Windows facts subsystem. Once we have that, the module becomes a lot easier (and the Windows facts can be reused in other parts of Ansible).

Commit:
Renamed files and module to win_disk_management

Commit:
Changed module options to fit Ansible guidelines

Commit:
Renamed module parameter options to comply with Ansible guidelines, adjusted PS file accordingly

Commit:
Renamed module parameter options to comply with Ansible guidelines, fixed one failure in code

marqelme commented:
@jhawkesworth
Now I have changed the module name to win_disk_management , because I also think the name fits better.
I have also changed the parameter options and logs etc. to comply with the Ansible naming guidelines.

@dagwieers 
Sure, if the disk selection part could be outsourced to the facts, this is also an option I would support you with.

Commit:
Adjusted options, parameters, switches and documentation to be more Ansible compliant, changed some other PS code

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,716 @@
+$params = Parse-Args $args;
+
+# Create a new result object
+$result = [pscustomobject]@{changed=$false;log=[pscustomobject]@{};general_log=[pscustomobject]@{};parameters=[pscustomobject]@{};switches=[pscustomobject]@{}}
```
We don't use `[pscustomobject]` anymore for results, just use a normal hashtable like below
```
$result = @{
    changed = $false
    log = @{}
    general_log = @{}
    ...
}
```

marqelme replied:
changed and implemented

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,716 @@
+
+## Extract each attributes into a variable
+# Find attributes
+$Size = Get-Attr -obj $params -name "disk_size" -resultobj $result -failifempty $true
```
Use `Get-AnsibleParam -obj $params -name "disk_size" -type "str/int/bool/list/path" -failifempty $true` instead of `Get-Attr`. Also you don't need to pass in `-resultobj $result` as it really does nothing.

marqelme replied:
changed and implemented

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,716 @@
+## Extract each attributes into a variable
+# Find attributes
+$Size = Get-Attr -obj $params -name "disk_size" -resultobj $result -failifempty $true
+$FindPartitionStyle = Get-Attr -obj $params -name "partition_style_select" -default "raw" -ValidateSet "raw","mbr","gpt" -resultobj $result -failifempty $false
```
Just omit `-failifempty $false` if the parameter is not mandatory.

marqelme replied:
changed and implemented

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,716 @@
+Invoke-Expression '"rescan" | diskpart' | Out-Null
+   }
+catch{
+Set-Attr $result.general_log "rescan_disks" "failed"
```
When moving away from `[pscustomobject]` with results, you no longer have to use `Set-Attr`. Just do `$result.general_log.rescan_disks = "failed"

marqelme replied:
changed and implemented

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,716 @@
+}
+catch{
+Set-Attr $result.general_log "search_disk" "failed"
+Fail-Json $result $_
```
Would be good to put some context around where this failed like `Fail-Json -obj $result -message "failed to get disk info for ...: $($_.Exception.Message)"`

marqelme replied:
changed and implemented

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,323 @@
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['stableinterface'],
```
status should be `preview`

marqelme replied:
changed and implemented

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,323 @@
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['stableinterface'],
+                    'supported_by': 'core'}
```
supported_by should be `community`

marqelme replied:
changed and implemented

jborean93 reviewed (changes requested):
Thanks for the PR, there are a few things that would need to be changed before this would be merged.
* new modules should support check mode
* new modules should come with integration tests, you can see existing examples under `test/integration/targets/win_*`
* diff mode is good to have but not mandatory
* the formatting should be consistent with our other modules, currently the indenting varies in the current PR and some functions aren't indented at all
* A lot of the cmdlets used in this aren't available with pre Server 2008 R2 and Windows 7, we would need to add this to the documentation
* It would be preferable to split the facts gathering and actions into separate module (like `win_disk_facts`), as stated in https://github.com/ansible/ansible/pull/27837, this probably doesn't warrant to be in the standard facts as the cmdlets aren't available in all Windows versions.

I've left some inline comments of areas where the standard practice has changed since past releases, I haven't added a comment for every instance but rather the first one I saw so you can see examples of what I mean for the list above.

Commit:
Improved documentation, options and parameters as suggested by community, made output more modular

marqelme commented:
Now improved the output and the documentation.
@jborean93 
Have already implemented some of your change requests (preview, community, note to for working Windows version in documentation, Get-Ansibleparam and adjusted types and parameter initialization).
I will have a look to your other points now and will try to implement.

Commit:
Adjusted py file

Commit:
Adjusted metadata information in .py file, added parameter types in .ps1 file

Commit:
Adjusted metadata version to 1.1 in .py file

Commit:
Adjusted metadata in .py file

Commit:
Adjusted metadata in .py file, removed customobject with hashtable in .ps1 file

Commit:
Added environment for test scenario

Commit:
Put context around Fail-Json ouput when modules fails because of errors; adjusted logging

Commit:
Fixed wrong typing

Commit:
Fixed syntax error

mattclay commented:
Closing and re-opening to trigger CI.

Commit:
Added --check_mode option in both files, added option read_only to select disk, improved documentation and code

Commit:
Removed trailing whitespace

marqelme commented:
**@jborean93** 

**-** have integrated check_mode now in PS script

**-** will add integration test next days

**-** diff_mode: no files will be changed in my module, so if it is not mandatory I will not integrate this

**-** have improved the formatting to be compliant with Ansible

**-** have added the hint to the documentation that the module can be used from Windows 6.3 on 

**-** to split the facts gathering and actions into separate module:

How to proceed with this? Can we use PR #27837 for it?
To add support for former Windows versions will be a lot of work, should I add it, is this mandatory?

jborean93 commented:
@marqelme looks good, I'll do another review once you have added the tests. You don't need to add diff support it isn't mandatory just a nice to have. For the facts I wouldn't add to the PR but rather create a new module calls `win_something_facts` and create a brand new PR. Just add the values you want to add for facts to the results object.

Commit:
Added main.yml content, added tests.yml for integration tests

Commit:
Fixed tests.yml

marqelme commented:
**@jborean93**

I have now added the integration test, but it fails with some error in the Shippable test.
Do you have an idea what is wrong in the code?

Commit:
Fixed tests.yml

Commit:
Added dfaults main.yml, adjusted tests.yml

Commit:
Adjusted syntax error in tests.yml

Commit:
Adjusted syntax error in tests.yml

Commit:
Adjusted errors in tests.yml

marqelme commented:
**@jborean93** 

Now the integration tests passed the shippable checks.
The two tests in tests.yml passes the checks because they fail since no disk is attached / could be found.

I can not implement a test where the test tries to select and manage/setup a disk in fact.
Because in the test scenarios there is no disk attached.
Naturally my module is only working if a disk is attached to a server otherwise it can't find and select a disk.
Is there any way to realize such a scenario in the integration tests?

Commit:
Adjusted errors in tests.yml

Commit:
Adjusted errors in tests.yml

Commit:
Adjusted errors in tests.yml

Commit:
Adjusted errors in tests.yml

Commit:
Adjusted errors in tests.yml

Commit:
Adjusted errors in tests.yml

Commit:
Adjusted errors in tests.yml

marqelme commented:
**Regarding the facts/selection in a separate module:**
Did I get it right, that with this module, you should be able to search for a disk and select the disk, then store this selected disk information in module result and use this in a variable to invoke win_disk_management with this variable?
Something like that playbook:

- name: select disk
  win_disk_select:
    disk_size: 100
    partition_style: raw
  register: selected_disk

- name: manage disk
  win_disk_management:
    disk: selected_disk

But then I still have to check in win_disk_management if the disk contains a volume or a partition.
Or this should be also a part of win_disk_select?
But then you have a lack in security in win_disk_management I think, because every disk would be used by it.

Commit:
Added option Number to select disk

Commit:
Added option Number to select disk

Commit:
corrected size option

Commit:
added convert disk log

Commit:
Added convert variable and fixed disk option

Commit:
fixed null valued expression error

Commit:
fixed whitespace error

Commit:
added convertion of variable logging

Commit:
added convertion of option variables logging

marqelme commented:
@jborean93 
Any updates from your side?

jborean93 commented:
@marqelme 

It is fine if you don't have extensive tests just do what you can. In this case it probably won't be possible to have a disk to test with so having some that test failures and that the script actually executes is probably good enough.

For the facts, I see the current iteration is returning a lot of facts based on the disk. Ultimately what happens is that someone will ask for a `state=query` option where they can use the module to gather facts about a particular component. While in some cases this was done we are trying to avoid going this route for a few reasons. What I am suggesting you do is

* limit the return values from this module to only what is important to know, e.g. don't list every attribute of a disk maybe just important details relating to a change
* (you don't have to do this) create a separate module that get's disk facts and adds them to the fact list, similar to how `gather_facts`, `setup` works.

To create a `*_facts` module all you need to do is create a module like the example below and populate the return results under an `ansible_facts` dict like so

```
#!powershell

#Requires -Module Ansible.ModuleUtils.Legacy

$result = @{
    changed = $false
    ansible_facts = @{
        ansible_disk = @(
            @{
                drive = "c"
                size = 10
        )
    }
}

Exit-Json $result
```

Let's say we call this module `win_disk_facts`, we can use it in a playbook/role like set

```
---
- name: get disk facts
  win_disk_facts:

- name: output first disk drive letter
  debug:
    var: ansible_facts.ansible_disk[0].drive
```

Anything under the `ansible_facts` dict is automatically registered for that host under `ansible_facts` so you don't need to mess around with `register` or `set_fact`. The benefits of this approach are

* Keeps the code and documentation for `win_disk_management` simpler as there is less going on and less paths to manage
* We don't need to add more code to `setup.ps1` which is already slow enough as it is
* People who want disk facts under setup just need to add an extra task to the begging to call the facts module and they achieve the same result

I've had another brief glance at your code and I'm still a bit concerned around the indentation in use. I see different indent lengths and other inconsistencies that would need to be fixed before merging in. Here is an example of a function with consistent indentation

```
Function Test-Function {
    Param(
        $string,
        $integer
    )

    if ($true) {
        Write-Host "consistent indentation"
    } else {
        Write-Host "on the same indent as the above"
    }

    $hashtable = @{
        key1 = "value1"
        key2 = "value2"
    }

    return "return value"
}
```

Feel free to use the C# style (as long as it is consistent) where braces are on a line of themselves like

```
if ($true)
{
    $true
}
else
{
    $false
}
```

Commit:
added when to integration test

Commit:
changed integration test

mattclay commented:
CI failure in integration test `test/integration/targets/win_disk_management/tasks/tests.yml:32`:

```
{
"failed": true, 
"msg": "The conditional check 'get_disk_actual.rc == 0' failed. The error was: error while evaluating conditional (get_disk_actual.rc == 0): 'dict object' has no attribute 'rc'"
}
```

Commit:
adjusted integration test

marqelme commented:
**@jborean93** 
Thanks for the feedback! 
Regarding the integration test I am fine now if you agree with it.
I will now decrease the facts and the return in the module to the minimal as you have suggested.
I would prefer to get this module merged here, after it I will create a new module and PR for gathering the facts as you have decribed above...
Regarding the consistent indentation you are right, I will correct this.

Is there a date when Ansible version 2.4 will be released and this module have to be merged at the latest?

jborean93 commented:
@marqelme, unfortunately it is too late for 2.4 so expect some CI failures around the documentation once the stable-2.4 branch is created. The freeze dates for 2.4 are here https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_4.rst. It will still be good to have a module to manage this though so 2.5 is looking more likely. Nonetheless once it is merged into devel that doesn't stop people from using it as they can just add it to their library path if need be before 2.5 is released.

marqelme commented:
**@jborean93** , ok great, thanks for this information!
Then we have time to get this merged to version 2.5

I will implement your proposals and suggestions as soon as possible.

Commit:
Adjusted integration test

Commit:
Adjusted examples

Commit:
Adjusted integration test

Commit:
Changed version to be added to 2.5

Commit:
Adjusted Number option query

Commit:
Adjusted number option description

Commit:
Added comments to examples

Commit:
Added comments to integration tests

Commit:
deleted whitespace

Commit:
Improved code

- improved indentation and code
- fixed bugs
- decreased disk facts

Commit:
Improved docu and examples

marqelme commented:
@jborean93 
I have adjusted indentation of code to be consistent and improved the whole module code.
Furthermore I have decreased the disk facts return in the module.
If the information about existing partitions and volumes is still to much let me know and I will reduce the output of it as well.
As already mentioned I will create a separate disk facts module once this module is ready.

Could you please check again whether module is now ready to get merged?

Commit:
Adjusted attributes

Commit:
Adjusted attributes

Commit:
Changed description

Commit:
Improved drive letter find and set

Commit:
Improved drive letter docu and main description

Commit:
Changed contact email address

marqelme commented:
**@gundalow**
Are the changes you have requested for the module are implemented?

Commit:
Changed documentation

Commit:
Added drive_letter in change_log docu

Commit:
Improved code, new skills

- Size is not mandatory
- Number option can be passed
- introduced logging option
- improved logging
- code improvements

Commit:
Small update

Commit:
Improved docu and logging

Commit:
Improved testing due to changes

Commit:
Removed whitespace

Commit:
Correcte query

jborean93 commented:
@marqelme I haven't forgotten about this, I am planning on looking at it sometime this week.

Commit:
Improved option value convertion

Commit:
Changed contact email, improved docu

Commit:
Replaced Out-Null by $null

Commit:
Improved docu

Commit:
Adjusted docu

Commit:
Added documentation items

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+# Find attributes
+$check_mode = Get-AnsibleParam -obj $params -name "_ansible_check_mode" -type "bool" -default $false
+$Logging = Get-AnsibleParam -obj $params -name "logging" -type "str" -default "standard" -ValidateSet "standard","verbose"
+$Size = Get-AnsibleParam -obj $params -name "size" -type "str/int"
```
Type can only have 1 value, if people should only set numbers then set this to `int` and Ansible will then convert a string to an int if necessary so you don't have to.

marqelme replied:
fixed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+$FindPartitionStyle = Get-AnsibleParam -obj $params -name "partition_style_select" -type "str" -default "raw" -ValidateSet "raw","mbr","gpt"
+$OperationalStatus = Get-AnsibleParam -obj $params -name "operational_status" -type "str" -default "offline" -ValidateSet "offline","online"
+$ReadOnly = Get-AnsibleParam -obj $params -name "read_only" -type "bool" -default $true
+$Number = Get-AnsibleParam -obj $params -name "number" -type "str/int"
```
ditto about `str/int`

marqelme replied:
fixed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+# Set attributes file system
+$FileSystem = Get-AnsibleParam -obj $params -name "file_system" -type "str" -default "ntfs" -ValidateSet "ntfs","refs"
+$Label = Get-AnsibleParam -obj $params -name "label" -type "str" -default "ansible_disk"
+$AllocUnitSize = Get-AnsibleParam -obj $params -name "allocation_unit_size" -type "str/int" -default 4 -ValidateSet 4,8,16,32,64
```
ditto about `str/int`

marqelme replied:
fixed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+#
+# Copyright 2017, Marc Tschapek <marc.tschapek@itelligencegroup.com>
+#
+# Ansible is free software: you can redistribute it and/or modify
```
We've recently made the copyright text shorter, have a look at the module development guides to see how to write the copyright text http://docs.ansible.com/ansible/2.4/dev_guide/developing_modules_general_windows.html#windows-new-module-development.

marqelme replied:
replaced

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+    }
+}
+$SizeInteger = $false
+if ($Size -ne $null) {
```
This whole block isn't needed if you set the `-type` of size to `int`. Will save you a lot of lines of code

marqelme replied:
removed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+    }
+}
+
+if (($AllocUnitSize.GetType()).Name -eq 'String') {
```
Same thing here, just use the `-type "int"` option and this whole thing is unecessary

marqelme replied:
removed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+}
+
+$NumberInteger = $false
+if ($Number -ne $null) {
```
Ditto

marqelme replied:
removed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+if ($DriveLetter -ne $null) {
+    if ($DriveLetter -like "[a-z]") {
+        try {
+            [char]$DriveLetter = [convert]::ToChar($DriveLetter)
```
I there a reason why you are converting it to a `[char]`, won't a string suffice and you don't need all this boilerplate code? You still have to validate that it is a single character but no casting would be required.

marqelme replied:
reason was that char is more space-saving but I I have removed this now

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+# Show option values
+if ($Logging -eq "verbose") {
+    $result += @{ 
+        option_values_passed = @{
```
I don't think this is really necessary, while there is a bug right now with Windows, the Ansible engine should be showing the invocation args independently from the module.

marqelme replied:
removed logging by diff_mode

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+                $Number
+        )
+
+        if ($DiskSizeInteger -and $DiskNumberInteger) {
```
No idea if `0` is a valid option, but if it is, the above will return `false`. You would have to change your checks to if ($DiskSizeInteger -ne $null -and $DiskNumberInteger -ne $null)`

marqelme replied:
0 was not possible in above query, but anyway I have replaced that because of other relations

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+                ($_.PartitionStyle -eq $PartitionStyle) -and ($_.OperationalStatus -eq $OperationalStatus) -and ($_.IsReadOnly -eq $ReadOnly) -and ($_.Size -eq $DiskSize) -and ($_.Number -eq $Number)
+            }
+
+            return $disk
```
Move `return $disk` outside the if branch so you only write it once

marqelme replied:
fixed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+        if (-not $Deactivate) {
+            $null = Set-Disk -Number ($Disk.Number) -IsOffline $false
+
+            return
```
`return` isn't required here

marqelme replied:
removed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+                [switch]$Deactivate
+        )
+
+        if (-not $Deactivate) {
```
Because you are passing a switch, this whole function can be simplified to

```
function Set-OperationalStatus {
    param(
        $Disk,
        [switch]$Deactivate
    )
    $null = Set-Disk -Number ($Disk.Number) -IsOffline $Deactivate.IsPresent
}
```

marqelme replied:
cool, thanks for the hint, fixed it

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+                [switch]$Deactivate
+        )
+
+        if (-not $Deactivate) {
```
See `Set-OperationalStatus` as how this can be simplified

marqelme replied:
fixed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+
+        $FoundVolume = Get-Volume | Where-Object {
+            $Partition.AccessPaths -like $_.ObjectId
+        }
```
Should we be validating that we found a volume, not sure if `$null` is a valid result from this.

marqelme replied:
Added $false as the return if no volume was found, I a volume will be found the volume object will be returned

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+
+        $null = $Disk| Initialize-Disk -PartitionStyle $PartitionStyle -Confirm:$false
+
+        return
```
`return` is not required

marqelme replied:
removed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+
+        $null = Invoke-Expression "'Select Disk $($Disk.Number)','Convert $($PartitionStyle)' | diskpart"
+
+        return
```
`return` is not required

marqelme replied:
removed

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+                    return $CheckService
+                }
+                default {
+                    throw "Neither '-Stop', '-Start' nor 'Check' switch was passed to the function when invoked, without the switches the service can not be maintained"
```
I would reword this to something like `Manage-ShellHWService -Action must be 'Stop', 'Start', or 'Check', was '$action'`

marqelme replied:
removed default{} because it will not be used in the module

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+
+        if ($FileSystem -eq "ntfs") {
+            if (-not $FileSystemLargeFRS) {
+                $CreatedVolume = $Volume | Format-Volume @ParaVol -ShortFileNameSupport $FileSystemShortNames -Force -Confirm:$false
```
Instead of calling `Format-Volume` 3 things, call it once and add the relevant switch to `$ParaVol` based on the FileSystem set, e.g.

```
$ParaVol = @{
    FileSystem = $FileSystem
    NewFileSystemLabel = $FileSystemLabel
    AllocationUnitSize = $Alloc
}

if ($FileSystem -eq "ntfs") {
    $ParaVol['ShortFileNameSupport'] = $FileSystemShortNames
    if ($FileSystemLargeFRS) {
        $ParaVol['UseLargeFRS'] = $true
    }
} elseif ($FileSystem -eq "refs") {
    $ParaVol['SetIntegrityStreams'] = $FileSystemIntegrityStream
}

$CreatedVolume = $Volume | Format-Volume @ParaVol -Force -Confirm:$false
```

marqelme replied:
thanks for the advice, I have added this

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+$ParamsDisk = @{
+    DiskSizeInteger = $SizeInteger
+    DiskNumberInteger = $NumberInteger
+    DiskSize = "$(if ($SizeInteger) {
```
Move this outside of this definition to make it easier to understand what is going wrong here

marqelme replied:
replaced it

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+if ($disk) {
+    $diskcount = $disk | Measure-Object | Select-Object  -ExpandProperty Count
+    if ($diskcount -ge 2) {
+        $disk = $disk[0]
```
Should you not be failing here instead of just picking the first disk?

marqelme replied:
no, because this will only happen if no number option was passed to the module and exactly the same disks were found with the select options. Then it doesn't matter which disk will be chosen and I select the first found.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,560 @@
+
+# (c) 2017, Marc Tschapek <marc.tschapek@itelligencegroup.com>
+#
+# This file is part of Ansible
```
This can be simplified like other documentation modules now.

marqelme replied:
replaced

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1411 @@
+
+# Create a new result object
+$result = @{
+    changed = $false
```
TBH with you, I don't think we should be adding all these options to the result. I think we should just be returning  `changed` and nothing else. The amount of code that is in this module to handle all this is quite large and makes it seem more complex than what it is.

If you really want logging options I would create a function that writes the log information to a file on the Windows both like what `win_updates` does and just keep on appending log info with datetime stamps.

As for the disk information stuff, keep it out as it should really be part of a facts/stat module instead and just causes more complication here.

marqelme replied:
removed a lot of logging, added diff_mode

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,560 @@
+module: win_disk_management
+version_added: '2.5'
+short_description: A Windows disk management module
+description:
```
A lot of the entries here should be in the description of each relevant option. This is for a high level overview of what the module does. The `notes` entry can also be used for info about the module if it doesn't fit in a option.

marqelme replied:
fixed that

jborean93 reviewed (changes requested):
Thanks @marqelme for the changes, I've gone through the first half of the modules and I've come to the conclusion that all the logging stuff you've added is just complicating things. We need to be able to maintain this module going forward if we are to include it within Ansible and at 1500 lines it becomes really hard.

I'm going to say that you should remove all the logging and disk info return values and just focus on the code. If you really want to have some sort of log, I would do what win_updates does and output to a file on the server but I don't think it is really necessary here. If it is a matter of helping you develop this module I would really encourage you to read http://docs.ansible.com/ansible/2.4/dev_guide/developing_modules_general_windows.html#windows-debugging as it helps you to run/debug the module on a Windows box. This makes it a whole lot easier to test and build modules as you don't need to run it through Ansible and changes can be made inline and breakpoints used to inspect variables at a point in time.

Commit:
Updated whole module

- Added diff_mode
- removed logs
- Code optimization

Commit:
Removed invalid check

Commit:
Updated documentation

Commit:
Corrected Size query

Commit:
Removed whitespace

marqelme commented:
**@jborean93** 
Thanks for all the input and suggestions. I have tried to realize all you have mentioned and remarked.
I have commented all your requests and already uploaded the updated code.
I have removed a lot of logging and almost 400 lines of code.
As an addition I have added the diff_mode to provide an opportunity to output the changes if a user wants it.
Could you please check the code again?

P.S. 
Will there be a way in future to validate the code in more detail than with ValidateSet?
Would be cool to have ValidatePattern, ValidateLength etc.
For instance I have now added all values for the drive letter (A-Z) to ValidateSet which is not very convenient and I am not sure if this is the right way.

Furthermore I have checked the size option of the module where it should be possible to pass an int64 value to. 
But I am not quite sure whether Ansible is converting it in the right way because I had some strange errors. An example would be to pass the value 2147483648 which is an int64. I always got a wrong value converted.
Could you please verify that?
Thank you!

One more question:
Is it ok that --check contains --diff?
Means that if you would use --check for the module it would give you what would be changed without extra passing --diff.
Or is that not the right way (--check would then need --diff to show what would be changed).

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+
+# (c) 2017, Marc Tschapek <marc.tschapek@itelligencegroup.com>
```
The format for the copyright (according to documentation) is:
```python
# Copyright: (c) 2017, Marc Tschapek <marc.tschapek@itelligencegroup.com>
```
(Not a showstopper)

marqelme replied:
replaced

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+---
+module: win_disk_management
+version_added: '2.5'
+short_description: A Windows disk management module
```
This does not add any information, maybe do something like:
```yaml
short_description: Manage disks, partitions and file systems
```

marqelme replied:
improved

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - Size of the disk in gigabyte which will be selected
+        - If a size is defined the module will try to select the disk with this passed size
+        - Size value must be equal or greater than 1gb
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+      description:
+        - Select disk option
+        - Partition style of the disk which will be selected
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+      description:
+        - Select disk option
+        - Operational Status of the disk which will be selected
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+      description:
+        - Select disk option
+        - Read-only status of the disk which will be selected (true=read-only,false=writeable)
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - If a number is defined the module will try to select the disk with this passed number
+        - Passed value will be checked in the beginning of the module whether it is an int32 value
+        - If it is of type in64 the module will be canceled
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+      description:
+        - Manage disk option
+        - Partition style which will be set on selected disk
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - If this drive letter is already in use on the target the module will be canceled
+        - If no drive_letter option value was defined the module will use a free drive letter on the target randomly
+        - If no free drive lettter is left on the target the module will be canceled
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - Maximum volume size for ntfs is 256gb, for refs 1208925819614650gb (1 yobibyte)
+        - If the disk size of the selected disk does not match with the passed value for option
+        - file_system (e.g. "ntfs" over 256000gb) the module will be canceled
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+      description:
+        - Manage disk option
+        - File system label which should be set for the file system on found disk
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - Manage disk option
+        - Allocation Unit size which will be set for the file system on selected disk (possible values for file system ntfs 4,8,16,32,64kb;refs 64kb)
+        - If option file_system is set to "refs" the allocation unit size will be automatically adjusted to "64" (kb)
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - Switch to set Large FRS option for file system on selected disk, solely settable for ntfs file system
+        - If this manage disk option was passed to the module it checks whether it is suitable with the passed file_system value
+        - If they do not fit this manage disk options will be automatically disabled
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - Switch to set Short Names option for file system on selected disk, solely settable for ntfs file system
+        - If this manage disk option was passed to the module it checks whether it is suitable with the passed file_system value
+        - If they do not fit this manage disk options will be automatically disabled
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+        - Switch to set Integrity Streams option for file system on selected disk, solely settable for refs file system
+        - If this manage disk option was passed to the module it checks whether it is suitable with the passed file_system value
+        - If they do not fit this manage disk options will be automatically disabled
+      required: false
```
If it's not required, you can leave out `required: false`.

marqelme replied:
removed

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+  - To identify the options which are used to manage a disk consider the "Manage disk option" hint in the option description.
+  - In order to find only one disk on the target you can use the size and/or number option for the search of the disk.
+  - If no size and number option value was defined and multiple disks were found on the target based on the passed option values
+  - the module will select the first disk found.
```
Here's something wrong. If you want to continue a line, you shouldn't start it with a dash.

marqelme replied:
changed to no dash

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+  - The module detects any existing volume and/or partition on the selected disk and will cancel the module in this case.
+  - If the disk is not yet initialized the module will initialize the disk (set partition style, online and writeable).
+  - If the disk is initialized already the script will try to set the disk to "online" and "writeable" (read-only eq. false)
+  - if it's not the state of the disk already.
```
Here's something wrong. If you want to continue a line, you shouldn't start it with a dash.

marqelme replied:
changed to no dash

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+  - Further in this case it will convert the partition style of the disk to the selected partition style if needed.
+  - The script will stop and start the service "ShellHWService" again in order to avoid disk management GUI messages.
+  - If the script fails with an error and the operational status was set from "offline" to "online" before the script will try to set the disk to
+  - operational status "offline" again but will not be canceled if set "offline" again fails.
```
Here's something wrong. If you want to continue a line, you shouldn't start it with a dash.

marqelme replied:
changed to no dash

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+  - If the script fails with an error and the operational status was set from "offline" to "online" before the script will try to set the disk to
+  - operational status "offline" again but will not be canceled if set "offline" again fails.
+  - If the script fails with an error and the writeable status was set from "read-only" to "writeable" before the script will try to set the disk to
+  - writeable status "read-only" again but will not be canceled if set "read-only" again fails.
```
Here's something wrong. If you want to continue a line, you shouldn't start it with a dash.

marqelme replied:
changed to no dash

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+  - operational status "offline" again but will not be canceled if set "offline" again fails.
+  - If the script fails with an error and the writeable status was set from "read-only" to "writeable" before the script will try to set the disk to
+  - writeable status "read-only" again but will not be canceled if set "read-only" again fails.
+  - If you use the _ansible_diff option you will get detailed information about the changes on the target.
```
This is not how it is used, the option used is `--diff`.

marqelme replied:
changed to --diff

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+  - If the script fails with an error and the writeable status was set from "read-only" to "writeable" before the script will try to set the disk to
+  - writeable status "read-only" again but will not be canceled if set "read-only" again fails.
+  - If you use the _ansible_diff option you will get detailed information about the changes on the target.
+  - If you use the _ansible_check_mode option nothing will be changed on the target but you will get information what would be changed.
```
This is not how it is used, the option used is `--check`.

marqelme replied:
changed to --check

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,244 @@
+options:
+  size:
+      description:
+        - Select disk option
```
So all the entries for descriptions should have a trailing dot. Only the short_description must not have a trailing dot. (To keep the index clean).

marqelme replied:
added trailing dots to all descriptions

Commit:
Corretec docu mistakes

Commit:
Corrected docu again

marqelme reviewed (commented):
Reviewed your requests and improved the docu

marqelme reviewed (commented):
Implemented your requests

marqelme reviewed (commented):
Implemented your requests

Commit:
Improved code

- Added drive letter L to validate set
- Improved Test-Path

Commit:
Fixed wrong Legacy module

Commit:
Improved code

Commit:
Improved docu

marqelme commented:
**@jborean93** **@dagwieers** 
Now I get the following error in sanity test:

The test ansible-test sanity --test validate-modules [?] failed with the following error:
`lib/ansible/modules/windows/win_disk_management.ps1:0:0: E211 Module #Requires should not end in .psm1: "Ansible.ModuleUtils.Legacy.psm1"`

Do yo know how to avoid this?

Commit:
Removed space

Commit:
Remove set_options result

jctanner commented:
bot_skip

mattclay commented:
@marqelme Use:

```powershell
#Requires -Module Ansible.ModuleUtils.Legacy
```

instead of:

```powershell
#Requires -Module Ansible.ModuleUtils.Legacy.psm1
```

mattclay commented:
@jctanner Why did you have the bot skip this PR?

jborean93 commented:
@marqelme this is a new check that was just recently added, remove the `.psm1` from the `#Requires -Module` statement and you should be fine.

Commit:
Removed Required module extension

Commit:
Improved docu

Commit:
Removed some typos

Commit:
Removed some typos

Commit:
Removed .psm1 extension

Commit:
Improved code

Removed refs "Check file system" for huge Size values because it is not necessary

Commit:
Adjusted Size and FileSystem docu

Commit:
Improved docu

Commit:
Improved docu

Commit:
Improved change_log

Commit:
Improved docu

Commit:
Code improvements

Commit:
Code improvements

Commit:
Updated code

- You can now choose between acccess_path or drive_letter or pass both to the module
- If no drive letter nor access_path is passed to the module a random drive letter will be chosen if one is available on the target

Commit:
Updated docu

Commit:
Removed whitespace

marqelme commented:
Would be cool if somebody could review the code and check for merge

marqelme commented:
@jborean93 
Added now new pull request regarding your suggested win_disk_facts module:
[https://github.com/ansible/ansible/pull/32935](url)

Commit:
Changed header

marqelme commented:
@dagwieers 
Can you trigger a review here please?

dagwieers commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,256 @@
+---
+module: win_disk_management
+version_added: '2.5'
+short_description: Select and manage a disk, it's partitions and file systems
```
Small typo: **its** instead of **it's**.

marqelme replied:
corrected it

dagwieers reviewed (approved):
LGTM

dagwieers commented:
@marqelme Not sure what you like me to do, instead of reviewing it again :-) But if you put it on the agenda of the next Windows WG meeting, we can discuss whatever is needed to get this merged (if it is not already ready to be merged right now, that is). cc: @jborean93 @nitzmahone @jhawkesworth @gundalow

Commit:
Corrected small typo

marqelme commented:
@dagwieers 
Thank you, added it to the agenda.

Commit:
Removed msg and changed from log docu

jborean93 commented:
Hey @marqelme, there is still too much happening formatting the return values, bearing in mind 2.5 freeze is coming up soon, I'm happy to merge this in if you don't mind me playing around with some of the stuff. I've had a few people ask me about this and it would be good to have in.

willthames commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,1136 @@
+}
+
+# Functions
+function Search-Disk {
```
This function doesn't really work the way I expect it to - it should cope with missing values far better. Really, if I just choose to pass `number`, that should be enough. Basically only compare if not null. I'm not sure how best to achieve this in powershell, I'm afraid.

marqelme replied:
Hi @willthames , it is enough if you pass a number or a size to the module. You can even pass no size or number to the module. The module will then try to find the disk with the information which is given.
The least information which can be given to select the disk is defined by the select options which have default values:
partition_style_select: raw
operational_status: offline
read_only: true

You only need to change them if you want to find a disk with different properties.
If you pass no size and number option it is possible that more than one disk will be found on the target.

Commit:
Removed needless logging

Commit:
Removed needless logging

marqelme commented:
Hey @jborean93 , I have removed 250 lines of needless code now.

Commit:
added requirement

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+    }
+}
+if ($Number -ne $null) {
+    if ([int32]2147483647 -lt $Number) {
```
You don't have to change it as this is more FYI but this can be set as `[Int32]::MaxValue` to make it clearer what it does.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+    }
+    [string]$DOperSt = $disk.OperationalStatus
+    [string]$DPartStyle = $disk.PartitionStyle
+    [string]$DROState = $disk.IsReadOnly
```
Keep this as a boolean so you can actually do boolean checks with `$true` and not `"True"`

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        Fail-Json -obj $result -message "No disk could be found and selected with the passed option values"
+}
+
+if ($check_mode -or $diff_mode) {
```
I think you should just remove this altogether, this isn't how diff_mode is implemented. Let's focus on getting the module in an acceptable state and we can look at diff mode in a later PR if needed.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+            try {
+                Set-OperationalStatus -Disk $disk
+            } catch {
+                if ($Verbose) {
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+}
+
+# Check volumes and partitions
+[string]$PartNumber = $disk.NumberOfPartitions
```
Why the cast to string?

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+#Requires -Module Ansible.ModuleUtils.Legacy
+$ErrorActionPreference = "Stop"
```
If you add `#AnsibleRequires -OSVersion 6.2` you don't need the `Get-Disk` check.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        if (-not $check_mode) {
+            # Set online
+            try {
+                Set-OperationalStatus -Disk $disk
```
You are calling this function regardless if the partition style of the disk is raw and it is already online causing it to return changed even when it wasn't. Idempotency is a bit part of Ansible and we want to ensure the modules follow this. I see this in a few more places down the track as well.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+                    try {
+                        Set-OperationalStatus -Disk $disk -Deactivate
+                    } catch {
+                        $OPStatusFailed = $true
```
Just have an empty catch block, changed would be whatever it is and shouldn't change regardless of this operation, i.e. you are effectively saying set changed to true when `$OpStatusFailed` as `$SetOnline` is true. `$SetOnline` is only set to true straight after you set `$result.changed = $true` a few lines above this. This behaviour is repeated a few times below as well.

You should also make the reverting actions into it's own function where you pass in `$SetOnline`, `$SetWritable` and try and revert the actions. This way you only need to call the revert function here (and below) instead of repeating yourself multiple times.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        Fail-Json -obj $result -message "General error while searching for volumes on the selected disk: $($_.Exception.Message)"
+    }
+    # Existent volumes and partitions
+    if (-not $volume) {
```
This doesn't make sense at all, in both if blocks you are ending with `Fail-Json`

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+            }
+            Fail-Json -obj $result -message "No free drive letter left on the target"
+        }
+} elseif ($DriveLetter -eq [string]) {
```
`-is [string]`

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+} elseif ($DriveLetter -eq [string]) {
+    # Use defined drive letter
+    try {
+        $CheckLetter = Get-ChildItem Function:[$DriveLetter]: | Foreach-Object {
```
s/Function:[$DriveLetter]:/-Path $DriveLetter/

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+
+# Check set option values
+# Check drive letter and access path
+if (($DriveLetter -eq $null) -and ($PartitionAccessPath -eq $null)) {
```
Can you move this whole section above where you set the drive online and writeable as you won't need the reverting actions at all.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+            Fail-Json -obj $result -message "The check to get free drive letters on the target failed"
+        }
+        if ($DriveLetter) {
+            $DriveLetter = $DriveLetter.TrimEnd(":")
```
Substring to get the first char is a better option here in case the return value ever changes in the future

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+    }
+} elseif ($FileSystem -eq "refs") {
+    if ($AllocUnitSize -ne 64) {
+        $AllocUnitSize = 64    
```
This should throw a failure as the unit size people specify is invalid with the refs file system.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+    }
+}
+# Check file system
+if ($FileSystem -eq "ntfs") {
```
Can you move above the set online and writeable so you don't need to revert the values

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+# Initialize / convert disk
+if ($DPartStyle -eq "RAW") {
+    if (-not $check_mode) {
+        if ($DOperSt -eq "Offline") {
```
I'm confused, I don't understand why you are setting these. From what I am seeing in the code, `$SetOnline` and `$SetWritable` are used when handling failures to revert actions, in this case the variables should already have been set accordingly if the module actually did the action and so these statements are not needed.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+            }
+            Fail-Json -obj $result -message "Failed to initialize the disk: $($_.Exception.Message)"
+        }
+        if ($diff_mode) {
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        }
+        $result.changed = $true
+    } else {
+        $result.change_log.initializing = "Disk with partition style $FindPartitionStyle (partition_style_select) will not be initialized to $SetPartitionStyle (partition_style_set) due to activated check_mode"
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+                Fail-Json -obj $result -message "Failed to convert the disk: $($_.Exception.Message)"
+            }
+            if ($diff_mode) {
+                $result.change_log.converting = "Partition style $FindPartitionStyle (partition_style_select) was converted to $SetPartitionStyle (partition_style_set)"
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+            }
+            $result.changed = $true
+        } else {
+            $result.change_log.converting = "Disk will not be converted from partition style $FindPartitionStyle (partition_style_select) to $SetPartitionStyle (partition_style_set) due to activated check_mode"
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        $null = Invoke-Expression "'Select Disk $($Disk.Number)','Convert $($PartitionStyle)' | diskpart"
+}
+
+function Manage-ShellHWService {
```
This should just be implemented where called using the cmdlet you want instead of a special function.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+$StartFailed = $false
+$CheckFailed = $false
+# Check ShellHWService
+try {
```
This all seems unnecessary, just do this instead

```
try {
    Stop-Service -Name .... -WhatIf:$check_mode
} catch {
    Fail-Json -obj $result -message "Failed to stop service ...: $($_.Exception.Message)"
}
```

Why are you stopping this service anyway?

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+}
+
+# Part disk
+if (-not $check_mode) {
```
If you truely do need to stop the service beforehand, put a `try {} finally { Start-Service -Name ... -WhatIf:$check_mode }` around all this so you don't need to have it in each failure scenario.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        }
+        if ($StopSuccess) {
+            try {
+                Manage-ShellHWService -Action "Start"
```
Just call `Start-Service -Name ... -WhatIf:$check_mode`, no use having a wrapper function around it.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+        }
+        Fail-Json -obj $result -message "Failed to create the partition on the disk: $($_.Exception.Message)"
+    }
+    if ($diff_mode) {
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+    }
+    $result.changed = $true
+} else {
+    $result.change_log.partitioning = "Disk will not be partitioned due to activated check_mode"
```
Remove

jborean93 commented on lib/ansible/modules/windows/win_disk_management.ps1:
```diff
@@ -0,0 +1,888 @@
+}
+
+# Finally check if ShellHWService needs to be started again
+if (-not $check_mode) {
```
```
if ($service_stopped) {
    try {
        Start-Service -Name ... -WhatIf:$check_mode
    } catch {
        Fail-Json -obj $result -message "failed to restart ... service: $($_.Exception.Message)"
    }
}
```

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,236 @@
+description:
+   - With the module you can select a disk on the target and manage it (e.g. initializing, partitioning, formatting) with several options.
+requirements:
+    - Windows 8.1 / Windows 2012R2 (NT 6.3) / PowerShell 4.0 or newer
```
Windows 8.1 / Windows 2012 (NT 6.2), no need to set the PS version as that will inevitably confuse people on older OS' who have upgraded to PS 4 or higher.

jborean93 commented on test/integration/targets/win_disk_management/tasks/main.yml:
```diff
@@ -0,0 +1,21 @@
+  failed_when: no
+  register: new_diskmanagement
+
+- name: Only run tests when Windows is capable
```
You should also run a failure test on OS' that are not supported to ensure a proper error message is returned.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,236 @@
+        - If no drive_letter option value was passed but a valid value for option access_path
+          was passed to the module it will will create this partition access path and no drive letter for
+          the partition on selected disk.
+        - If no drive_letter option value and no value for option access_path was passed to the module,
```
I think this should be a mandatory option when creating a disk and we don't randomly choose one.

jborean93 commented on lib/ansible/modules/windows/win_disk_management.py:
```diff
@@ -0,0 +1,236 @@
+      description:
+        - Manage disk option.
+        - Allocation unit size which will be set for the file system on selected disk (possible values for file system ntfs 4,8,16,32,64kb;refs 64kb).
+        - If option file_system is set to "refs" the allocation unit size will be automatically adjusted to "64" (kb).
```
This shouldn't be automatically adjusted, just a note saying refs style must be 64.

jborean93 reviewed (changes requested):
Hey @marqelme, everytime I come to review this module I keep on coming away with the feeling that it is doing too much and there is a lot of magic happening in the background. I still left some comments on this pass of the review but I think we should go back to the basics and relook at what we want from this module.

According to the description, this module does

* disk initialisation
* disk partitioning
* disk formatting

Basic Ansible philosophy when it comes to modules is for it to do one task and do it well. While this is not the case everywhere, a lot of the modules try to keep things at a minimum. You can't do this all the time but it is what we strive for. What happens when a module does too many things or interacts with too many components is

* It gets overly complex as the code to handle this just grows
* The relationship between the options is based around the action the user wants to do meaning there is a lot of unused options in each scenario
* You have multiple options for the same thing, e.g. `partition_style_select` and `partition_style_set` to select a partition and then change it, this should just be `partition_style` to ensure the state of the partition.
* In this case, you are trying to deal with disk formatting on a disk object when in reality this is on a partition level further complication the logic around what to do in the module

What we should really have instead is 3 modules

* `win_disk_management`
* `win_disk_partition`
* `win_disk_format`

`win_disk_management` would be used to initialise disks, where you can specify the volume to initialise and to what partition style. It would look something like

```
win_disk_management:
  disk: 1  # the disk ID (must be set as I don't know a way of uniquely identifying a disk)
  partition: mbr # the type of partition to initialise as
  state: online # can be online/offline
  type: basic  # can be basic/dynamic
```

I'm not sure how you could go about converting the disk once a partition is already there but maybe throwing an error is the best case.

`win_disk_partition` is used to manage the partitions of an initialised disk, would look like

```
win_disk_partition:
  disk: 1  # the disk ID where the partition is located  (must be set)
  size: ....  # the size in bytes of the partition to create
  mount: D  # can be not defined or to an NTFS folder (mount point)
  state: present  # present/absent
  active: yes  # yes/no
```

There could be more options but in the end you would select the partition by the disk ID and mount if specific and is in use. If the mount does not match and is not in use you would use the empty space left on the disk.

`win_disk_format` is used to format created partitions, would look like

```
win_disk_format:
  disk: 1  # used with partition instead of mount to select the partition
  partition: 1  # used with disk instead of mount to select the partition
  mount: D  # used instead of disk/partition being set
  allocation_unit: 4  # can not be specific and use the file_system default
  file_system: ntfs  # ntfs/refs/fat32
  quick: yes  # quick format yes/no
  compression: no  # enable file/folder compression NTFS only
```

The downside is that someone now has to has 3 tasks to convert a raw uninitialised disk to a useable formatted partition but the advantages are;

* The modules are simpler as they only deal with the components they touch, no reverting if partitioning failed and we initialised a disk
* When the modules are simpler it is a lot easier to debug and add new features for
* The options for the modules are easier to understand, no need to set `partition_style_select` and `partition_style_set` if you just want to change the disk 
partition
* It is a lot easier to review each individual component in a PR as the components the module interacts with is a lot smaller.

I know you've put a lot of time into this but now that some of the layers have been peeled away I can understand more of what is happening and that feeling of this is doing too much keeps on growing.

chopraaa commented:
Is this module still in the pipeline or are we going for `win_volume` and `win_partition`?

marqelme commented:
@jborean93 
That's something I also wanted to know...

jborean93 commented:
I believe it was decided in one of the working group meeting to split it up into 3 different modules.

marqelme commented:
Ok, if I got the time I will convert this into 3 independent modules.

Geogboe commented:
Hi All, I wanted to create a PR for a new module called win_volume to create, partition and format volumes on a given disk. This PR seems to indicate something like that is in the works but I can't find any reference to it. Anyone know where it's being worked on if at all?

marqelme commented:
Hi George

I have worked in it, but got no time to continue.
It was not merged because the Community wanted to split it in separate modules (one for creating partitions, one for volumes etc.)

Regards
Marc

Outlook für iOS<https://aka.ms/o0ukef> beziehen

________________________________
Von: George.iO <notifications@github.com>
Gesendet: Donnerstag, August 16, 2018 1:12 AM
An: ansible/ansible
Cc: Marc Tschapek; Mention
Betreff: Re: [ansible/ansible] Adding new Windows module win_disk_management (#27634)


Hi All, I wanted to create a PR for a new module called win_volume to create, partition and format volumes on a given disk. This PR seems to indicate something like that is in the works but I can't find any reference to it. Anyone know where it's being worked on if at all?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/ansible/ansible/pull/27634#issuecomment-413368255>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AcZinG2DWPapIoKtiTcb-N2a3Sb02KYZks5uRKragaJpZM4Oq_Jb>.

dagwieers commented:
Beware that the deadline for getting new features/modules accepted in [Ansible v2.7 is nearing](https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/roadmap/ROADMAP_2_7.rst#release-schedule), it is set to either 2018-08-23 or 2018-08-30. If you are blocked, or you need feedback, please discuss on IRC channel #ansible-windows or add a comment to the [Windows Working Group meeting agenda](https://github.com/ansible/community/issues/294) to get it resolved.

Geogboe commented:
@marqelme Thanks for update! @dagwieers I've never made an Ansible module before so I don't think I'll make that deadline :) but I'm happy to give it a shot for version 2.8!

dagwieers commented:
During the [next Windows WG meeting](https://github.com/ansible/community/issues/420) on 2019-01-29 we will review and decide on the inclusion of this module in Ansible v2.8. Your presence in the meeting could be helpful.

*If there's anything you wanted to improve/fix, please do so before the meeting.*

jborean93 commented:
We discussed this at the last Windows Working Group meeting and we agreed on the current position that this module should be split up into 3 separate modules;

* win_volume
* win_partition
* win_format

The win_partition module has already been done and are happy to see contributions to the other 2. We will close this PR as it stands, thanks to everyone involved for working on this one.
